src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -40,75 +45,57 @@ public RouteValueDictionary()
         /// Only public instance non-index properties are considered.
         /// </remarks>
         public RouteValueDictionary(object values)
+            : this()
         {
-            var dictionary = values as RouteValueDictionary;
-            if (dictionary != null)
+            if (values is RouteValueDictionary dictionary)
             {
-                var listStorage = dictionary._storage as ListStorage;
-                if (listStorage != null)
+                if (dictionary._propertyStorage != null)
                 {
-                    _storage = new ListStorage(listStorage);
+                    // PropertyStorage is immutable so we can just copy it.
+                    _propertyStorage = dictionary._propertyStorage;
+                    _count = dictionary._count;
```

Comment:
How worried are we about thread safety in RVD? i.e. someone modifying the other dictionary in-between _propertyStore being changed and _count being changed.

Reply:
Not worried at all 😆 Virtually everything in ASP.NET is thread-unsafe